packages/flutter_tools/lib/src/build_system/build_system.dart

```diff
@@ -0,0 +1,581 @@
+    Directory buildDir,
+    Directory cacheDir,
+    Directory copyDir,
+    TargetPlatform targetPlatform,
+    BuildMode buildMode,
+  }) {
+    assert(projectDir != null);
+    return Environment._(
+      projectDir: projectDir,
+      stampDir: stampDir ?? projectDir.childDirectory('build'),
+      buildDir: buildDir ?? projectDir.childDirectory('build'),
+      cacheDir: cacheDir ?? Cache.instance.getCacheArtifacts().childDirectory('engine'),
+      copyDir: copyDir ?? projectDir
+        .childDirectory(getHostFolderForTargetPlaltform(targetPlatform))
+        .childDirectory('flutter'),
+      targetPlatform: targetPlatform,
+      buildMode: buildMode,
+    );
+  }
+
+  Environment._({
+    @required this.projectDir,
+    @required this.stampDir,
+    @required this.buildDir,
+    @required this.cacheDir,
+    @required this.copyDir,
+    @required this.targetPlatform,
+    @required this.buildMode,
+  });
+
+  /// The `PROJECT_DIR` magic environment varaible.
+  final Directory projectDir;
+
+  /// The `STAMP_DIR` magic environment variable.
+  ///
+  /// Defaults to `{PROJECT_ROOT}/build`.
+  final Directory stampDir;
+
+  /// The `BUILD_DIR` magic environment variable.
+  ///
+  /// Defaults to `{PROJECT_ROOT}/build`.
+  final Directory buildDir;
+
+  /// The `CACHE_DIR` magic environment variable.
+  ///
+  /// Defaults to `{FLUTTER_ROOT}/bin/cache`.
+  final Directory cacheDir;
+
+  /// The `COPY_DIR` magic environment variable.
+  ///
+  /// Defaults to `{PROJECT_ROOT}/{host_folder}/flutter`
+  final Directory copyDir;
+
+  /// The currently selected build mode.
+  final BuildMode buildMode;
+
+  /// The current target platform, or `null` if none.
+  final TargetPlatform targetPlatform;
+}
+
+class BuildSystem {
+  const BuildSystem([this.targets = allTargets]);
+
+  final List<Target> targets;
+
+  /// Build the target `name` and all of its dependencies.
+  Future<void> build(
+    String name,
+    Environment environment,
+  ) async {
+    // Initialize any destination directories that don't currently exist.
+    if (!environment.cacheDir.existsSync()) {
+      environment.cacheDir.createSync(recursive: true);
+    }
+    if (!environment.copyDir.existsSync()) {
+      environment.copyDir.createSync(recursive: true);
+    }
+
+    // Compute the required order of targets.
+    final List<Target> ordered = _computeTargetOrder(targets, name, environment);
+
+    // Visit each target and check if its stamp is up to date. If so,
+    // we can safely skip it. Otherwise, we must invoke the associated rule.
+    for (Target target in ordered) {
+      final List<FileSystemEntity> inputs = target.resolveInputs(environment);
+      final _InvocationEvaluation evaluation = target._canSkipInvocation(inputs, environment);
+
+      if (evaluation.canSkip) {
+        printTrace('Skipping target: ${target.name}');
+        continue;
+      }
+      printTrace('${target.name}: Starting');
+      await target.invocation(inputs, environment);
```

Comment:
I think it would be good to consider designing around running build steps in parallel from the beginning. A possible strategy: Visit the dependency graph, at each target you recursively visit each dependency and get back a Future. When all of those Futures complete, run the Target's action on a Pool (from package:pool), wrapped in an AsyncMemoizer (from package:async) to avoid running targets more than once.